// @material-ui/pickers/makePickerWithState-00818f63.js
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/esm/objectWithoutProperties';
import React__default, { createElement, useMemo, useRef, useEffect, forwardRef } from 'react';
import { node, bool, object, func } from 'prop-types';
import { D as DIALOG_WIDTH, b as DIALOG_WIDTH_WIDER, u as useUtils, I as IS_TOUCH_DEVICE_MEDIA, M as MuiPickersAdapterContext } from './dimensions-91c32fe6.js';
import clsx from 'clsx';
import _extends from '@babel/runtime/helpers/esm/extends';
import { makeStyles } from '@material-ui/core/styles';
import { o as onSpaceOrEnter, e as executeInTheNextEventLoopTick } from './utils-ef34bcd7.js';
import { e as getDisplayDate, f as getTextFieldAriaText, K as KeyboardDateInput$1, p as parsePickerInputValue, P as Picker } from './Picker-019ee379.js';
import useMediaQuery from '@material-ui/core/useMediaQuery';
import Button from '@material-ui/core/Button';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import Dialog from '@material-ui/core/Dialog';
import { W as WrapperVariantContext, I as IsStaticVariantContext } from './WrapperVariantContext-a0f5915b.js';
import Popover from '@material-ui/core/Popover';
import Grow from '@material-ui/core/Grow';
import Paper from '@material-ui/core/Paper';
import TrapFocus from '@material-ui/core/Modal/TrapFocus';
import Popper from '@material-ui/core/Popper';
import { a as useGlobalKeyDown, k as keycode } from './useKeyDown-9f8b37a0.js';
import { u as usePickerState } from './usePickerState-1c348703.js';

var useStyles = makeStyles({
  dialogRoot: {
    minWidth: DIALOG_WIDTH
  },
  dialogRootWider: {
    minWidth: DIALOG_WIDTH_WIDER
  },
  dialogContainer: {
    '&:focus > $dialogRoot': {
      outline: 'auto',
      '@media (pointer:coarse)': {
        outline: 0
      }
    }
  },
  dialog: {
    '&:first-child': {
      padding: 0
    }
  },
  withAdditionalAction: {
    // set justifyContent to default value to fix IE11 layout bug
    // see https://github.com/mui-org/material-ui-pickers/pull/267
    justifyContent: 'flex-start',
    '& > *:first-child': {
      marginRight: 'auto'
    }
  }
}, {
  name: 'MuiPickersModal'
});
var ModalDialog = function ModalDialog(_ref) {
  var children = _ref.children,
      onAccept = _ref.onAccept,
      onDismiss = _ref.onDismiss,
      onClear = _ref.onClear,
      onSetToday = _ref.onSetToday,
      okLabel = _ref.okLabel,
      cancelLabel = _ref.cancelLabel,
      clearLabel = _ref.clearLabel,
      todayLabel = _ref.todayLabel,
      clearable = _ref.clearable,
      showTodayButton = _ref.showTodayButton,
      showTabs = _ref.showTabs,
      wider = _ref.wider,
      other = _objectWithoutProperties(_ref, ["children", "onAccept", "onDismiss", "onClear", "onSetToday", "okLabel", "cancelLabel", "clearLabel", "todayLabel", "clearable", "showTodayButton", "showTabs", "wider"]);

  var classes = useStyles();
  return createElement(Dialog, _extends({
    onClose: onDismiss,
    classes: {
      container: classes.dialogContainer,
      paper: clsx(classes.dialogRoot, wider && classes.dialogRootWider)
    }
  }, other), createElement(DialogContent, {
    children: children,
    className: classes.dialog
  }), createElement(DialogActions, {
    className: clsx((clearable || showTodayButton) && classes.withAdditionalAction)
  }, clearable && createElement(Button, {
    "data-mui-test": "clear-action-button",
    color: "primary",
    onClick: onClear
  }, clearLabel), showTodayButton && createElement(Button, {
    "data-mui-test": "today-action-button",
    color: "primary",
    onClick: onSetToday
  }, todayLabel), cancelLabel && createElement(Button, {
    color: "primary",
    onClick: onDismiss
  }, cancelLabel), okLabel && createElement(Button, {
    color: "primary",
    onClick: onAccept
  }, okLabel)));
};
ModalDialog.displayName = 'ModalDialog';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }
var PureDateInput = function PureDateInput(_ref) {
  var inputFormat = _ref.inputFormat,
      rawValue = _ref.rawValue,
      validationError = _ref.validationError,
      InputProps = _ref.InputProps,
      onOpen = _ref.openPicker,
      renderInput = _ref.renderInput,
      forwardedRef = _ref.forwardedRef,
      containerRef = _ref.containerRef,
      _ref$getOpenDialogAri = _ref.getOpenDialogAriaText,
      getOpenDialogAriaText = _ref$getOpenDialogAri === void 0 ? getTextFieldAriaText : _ref$getOpenDialogAri,
      disabled = _ref.disabled,
      label = _ref.label,
      _ref$TextFieldProps = _ref.TextFieldProps,
      TextFieldProps = _ref$TextFieldProps === void 0 ? {} : _ref$TextFieldProps;
  var utils = useUtils();
  var PureDateInputProps = useMemo(function () {
    return _objectSpread({}, InputProps, {
      readOnly: true
    });
  }, [InputProps]);
  var inputValue = getDisplayDate(utils, rawValue, inputFormat);
  return renderInput(_objectSpread({
    label: label,
    disabled: disabled,
    ref: containerRef,
    inputRef: forwardedRef,
    error: validationError,
    'aria-label': getOpenDialogAriaText(rawValue, utils),
    onClick: onOpen,
    value: inputValue,
    InputProps: PureDateInputProps,
    onKeyDown: onSpaceOrEnter(onOpen)
  }, TextFieldProps));
};
PureDateInput.displayName = 'PureDateInput';

var MobileWrapper = function MobileWrapper(_ref) {
  var open = _ref.open,
      children = _ref.children,
      okLabel = _ref.okLabel,
      cancelLabel = _ref.cancelLabel,
      clearLabel = _ref.clearLabel,
      todayLabel = _ref.todayLabel,
      showTodayButton = _ref.showTodayButton,
      clearable = _ref.clearable,
      DialogProps = _ref.DialogProps,
      showTabs = _ref.showTabs,
      wider = _ref.wider,
      DateInputProps = _ref.DateInputProps,
      onClear = _ref.onClear,
      onAccept = _ref.onAccept,
      onDismiss = _ref.onDismiss,
      onSetToday = _ref.onSetToday,
      PopoverProps = _ref.PopoverProps,
      displayStaticWrapperAs = _ref.displayStaticWrapperAs,
      KeyboardDateInputComponent = _ref.KeyboardDateInputComponent,
      _ref$PureDateInputCom = _ref.PureDateInputComponent,
      PureDateInputComponent = _ref$PureDateInputCom === void 0 ? PureDateInput : _ref$PureDateInputCom,
      other = _objectWithoutProperties(_ref, ["open", "children", "okLabel", "cancelLabel", "clearLabel", "todayLabel", "showTodayButton", "clearable", "DialogProps", "showTabs", "wider", "DateInputProps", "onClear", "onAccept", "onDismiss", "onSetToday", "PopoverProps", "displayStaticWrapperAs", "KeyboardDateInputComponent", "PureDateInputComponent"]);

  return createElement(WrapperVariantContext.Provider, {
    value: "mobile"
  }, createElement(PureDateInputComponent, _extends({}, other, DateInputProps)), createElement(ModalDialog, _extends({
    wider: wider,
    showTabs: showTabs,
    open: open,
    onClear: onClear,
    onAccept: onAccept,
    onDismiss: onDismiss,
    onSetToday: onSetToday,
    clearLabel: clearLabel,
    todayLabel: todayLabel,
    okLabel: okLabel,
    cancelLabel: cancelLabel,
    clearable: clearable,
    showTodayButton: showTodayButton,
    children: children,
    "data-mui-test": "mobile-wrapper-dialog"
  }, DialogProps)));
};
process.env.NODE_ENV !== "production" ? MobileWrapper.propTypes = {
  okLabel: node,
  cancelLabel: node,
  clearLabel: node,
  clearable: bool,
  todayLabel: node,
  showTodayButton: bool,
  DialogProps: object
} : void 0;
MobileWrapper.defaultProps = {
  okLabel: 'OK',
  cancelLabel: 'Cancel',
  clearLabel: 'Clear',
  todayLabel: 'Today',
  clearable: false,
  showTodayButton: false
};

var useStyles$1 = makeStyles({
  popover: {
    '&:focus': _defineProperty({
      outline: 'auto'
    }, IS_TOUCH_DEVICE_MEDIA, {
      outline: 0
    })
  }
});
var DesktopWrapper = function DesktopWrapper(_ref) {
  var open = _ref.open,
      children = _ref.children,
      PopoverProps = _ref.PopoverProps,
      onDismiss = _ref.onDismiss,
      DateInputProps = _ref.DateInputProps,
      _ref$KeyboardDateInpu = _ref.KeyboardDateInputComponent,
      KeyboardDateInputComponent = _ref$KeyboardDateInpu === void 0 ? KeyboardDateInput$1 : _ref$KeyboardDateInpu;
  var ref = useRef(null);
  var classes = useStyles$1();
  return createElement(WrapperVariantContext.Provider, {
    value: "desktop"
  }, createElement(KeyboardDateInputComponent, _extends({}, DateInputProps, {
    containerRef: ref
  })), createElement(Popover, _extends({
    role: "dialog",
    open: open,
    onClose: onDismiss,
    anchorEl: ref.current,
    classes: {
      paper: classes.popover
    },
    anchorOrigin: {
      vertical: 'bottom',
      horizontal: 'center'
    },
    transformOrigin: {
      vertical: 'top',
      horizontal: 'center'
    }
  }, PopoverProps), children));
};
process.env.NODE_ENV !== "production" ? DesktopWrapper.propTypes = {
  onOpen: func,
  onClose: func,
  PopoverProps: object
} : void 0;

var useStyles$2 = makeStyles(function (theme) {
  return {
    popper: {
      zIndex: theme.zIndex.modal
    },
    paper: {
      transformOrigin: 'top center',
      '&:focus': _defineProperty({
        outline: 'auto'
      }, IS_TOUCH_DEVICE_MEDIA, {
        outline: 0
      })
    },
    topTransition: {
      transformOrigin: 'bottom center'
    }
  };
});
var DesktopPopperWrapper = function DesktopPopperWrapper(_ref) {
  var open = _ref.open,
      children = _ref.children,
      PopperProps = _ref.PopperProps,
      onDismiss = _ref.onDismiss,
      DateInputProps = _ref.DateInputProps,
      _ref$TransitionCompon = _ref.TransitionComponent,
      TransitionComponent = _ref$TransitionCompon === void 0 ? Grow : _ref$TransitionCompon,
      _ref$KeyboardDateInpu = _ref.KeyboardDateInputComponent,
      KeyboardDateInputComponent = _ref$KeyboardDateInpu === void 0 ? KeyboardDateInput$1 : _ref$KeyboardDateInpu;
  var classes = useStyles$2();
  var inputRef = useRef(null);
  var popperRef = useRef(null);
  useGlobalKeyDown(open, _defineProperty({}, keycode.Esc, onDismiss));

  var handleBlur = function handleBlur() {
    executeInTheNextEventLoopTick(function () {
      var _inputRef$current, _popperRef$current;

      if (((_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.contains(document.activeElement)) || ((_popperRef$current = popperRef.current) === null || _popperRef$current === void 0 ? void 0 : _popperRef$current.contains(document.activeElement))) {
        return;
      }

      onDismiss();
    });
  };

  return createElement(WrapperVariantContext.Provider, {
    value: "desktop"
  }, createElement(KeyboardDateInputComponent, _extends({}, DateInputProps, {
    containerRef: inputRef,
    onBlur: handleBlur
  })), createElement(Popper, _extends({
    transition: true,
    placement: "bottom",
    open: open,
    anchorEl: inputRef.current
  }, PopperProps, {
    className: clsx(classes.popper, PopperProps === null || PopperProps === void 0 ? void 0 : PopperProps.className)
  }), function (_ref2) {
    var TransitionProps = _ref2.TransitionProps,
        placement = _ref2.placement;
    return createElement(TrapFocus, {
      open: open,
      disableAutoFocus: true,
      disableEnforceFocus: true,
      isEnabled: function isEnabled() {
        return true;
      },
      getDoc: function getDoc() {
        var _ref3, _popperRef$current2;

        return (_ref3 = (_popperRef$current2 = popperRef.current) === null || _popperRef$current2 === void 0 ? void 0 : _popperRef$current2.ownerDocument) !== null && _ref3 !== void 0 ? _ref3 : document;
      }
    }, createElement(TransitionComponent, _extends({}, TransitionProps, {
      timeout: 350
    }), createElement(Paper, {
      ref: popperRef,
      onBlur: handleBlur,
      tabIndex: -1,
      elevation: 8,
      className: clsx(classes.paper, placement === 'top' && classes.topTransition)
    }, children)));
  }));
};

var makeResponsiveWrapper = function makeResponsiveWrapper(DesktopWrapperComponent, MobileWrapperComponent) {
  var ResponsiveWrapper = function ResponsiveWrapper(_ref) {
    var _ref$desktopModeMedia = _ref.desktopModeMediaQuery,
        desktopModeMediaQuery = _ref$desktopModeMedia === void 0 ? IS_TOUCH_DEVICE_MEDIA : _ref$desktopModeMedia,
        okLabel = _ref.okLabel,
        cancelLabel = _ref.cancelLabel,
        clearLabel = _ref.clearLabel,
        todayLabel = _ref.todayLabel,
        showTodayButton = _ref.showTodayButton,
        clearable = _ref.clearable,
        DialogProps = _ref.DialogProps,
        PopoverProps = _ref.PopoverProps,
        PopperProps = _ref.PopperProps,
        TransitionComponent = _ref.TransitionComponent,
        displayStaticWrapperAs = _ref.displayStaticWrapperAs,
        other = _objectWithoutProperties(_ref, ["desktopModeMediaQuery", "okLabel", "cancelLabel", "clearLabel", "todayLabel", "showTodayButton", "clearable", "DialogProps", "PopoverProps", "PopperProps", "TransitionComponent", "displayStaticWrapperAs"]);

    var isDesktop = useMediaQuery(desktopModeMediaQuery);
    return isDesktop ? createElement(DesktopWrapperComponent, _extends({
      PopperProps: PopperProps,
      TransitionComponent: TransitionComponent,
      PopoverProps: PopoverProps
    }, other)) : createElement(MobileWrapperComponent, _extends({
      okLabel: okLabel,
      cancelLabel: cancelLabel,
      clearLabel: clearLabel,
      todayLabel: todayLabel,
      showTodayButton: showTodayButton,
      clearable: clearable,
      DialogProps: DialogProps
    }, other));
  };

  return ResponsiveWrapper;
};
var ResponsiveWrapper = makeResponsiveWrapper(DesktopWrapper, MobileWrapper);
var ResponsivePopperWrapper = makeResponsiveWrapper(DesktopPopperWrapper, MobileWrapper);

var useStyles$3 = makeStyles(function (theme) {
  return {
    staticWrapperRoot: {
      overflow: 'hidden',
      minWidth: DIALOG_WIDTH,
      display: 'flex',
      flexDirection: 'column',
      backgroundColor: theme.palette.background.paper
    }
  };
}, {
  name: 'MuiPickersStaticWrapper'
});
var StaticWrapper = function StaticWrapper(_ref) {
  var _ref$displayStaticWra = _ref.displayStaticWrapperAs,
      displayStaticWrapperAs = _ref$displayStaticWra === void 0 ? 'static' : _ref$displayStaticWra,
      children = _ref.children;
  var classes = useStyles$3();
  return createElement(IsStaticVariantContext.Provider, {
    value: true
  }, createElement(WrapperVariantContext.Provider, {
    value: displayStaticWrapperAs
  }, createElement("div", {
    className: classes.staticWrapperRoot,
    children: children
  })));
};

var defaultIsSameError = function defaultIsSameError(a, b) {
  return a === b;
};

function makeValidationHook(validateFn) {
  var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      defaultValidationError = _ref.defaultValidationError,
      _ref$isSameError = _ref.isSameError,
      isSameError = _ref$isSameError === void 0 ? defaultIsSameError : _ref$isSameError;

  return function (value, props) {
    var utils = useUtils();
    var previousValidationErrorRef = useRef(defaultValidationError || null);
    var validationError = validateFn(utils, value, props);
    useEffect(function () {
      if (props.onError && !isSameError(validationError, previousValidationErrorRef.current)) {
        props.onError(validationError, value);
      }

      previousValidationErrorRef.current = validationError;
    }, [previousValidationErrorRef, props, validationError, value]);
    return validationError;
  };
}

function withDateAdapterProp(Component) {
  return function (_ref) {
    var dateAdapter = _ref.dateAdapter,
        other = _objectWithoutProperties(_ref, ["dateAdapter"]);

    if (dateAdapter) {
      return createElement(MuiPickersAdapterContext.Provider, {
        value: dateAdapter
      }, createElement(Component, other));
    }

    return createElement(Component, other);
  };
}

/** Creates a component that rendering modal/popover/nothing and spreading props down to text field */
function makeWrapperComponent(Wrapper, _ref) {
  var KeyboardDateInputComponent = _ref.KeyboardDateInputComponent,
      PureDateInputComponent = _ref.PureDateInputComponent;

  function WrapperComponent(props) {
    var open = props.open,
        value = props.value,
        autoOk = props.autoOk,
        inputFormat = props.inputFormat,
        defaultHighlight = props.defaultHighlight,
        onChange = props.onChange,
        children = props.children,
        clearable = props.clearable,
        clearLabel = props.clearLabel,
        DialogProps = props.DialogProps,
        PopoverProps = props.PopoverProps,
        okLabel = props.okLabel,
        cancelLabel = props.cancelLabel,
        todayLabel = props.todayLabel,
        DateInputProps = props.DateInputProps,
        wrapperProps = props.wrapperProps,
        wider = props.wider,
        showTabs = props.showTabs,
        onAccept = props.onAccept,
        onClose = props.onClose,
        onOpen = props.onOpen,
        displayStaticWrapperAs = props.displayStaticWrapperAs,
        restPropsForTextField = _objectWithoutProperties(props, ["open", "value", "autoOk", "inputFormat", "defaultHighlight", "onChange", "children", "clearable", "clearLabel", "DialogProps", "PopoverProps", "okLabel", "cancelLabel", "todayLabel", "DateInputProps", "wrapperProps", "wider", "showTabs", "onAccept", "onClose", "onOpen", "displayStaticWrapperAs"]);

    var WrapperComponent = Wrapper;
    return React__default.createElement(WrapperComponent, _extends({
      clearable: clearable,
      clearLabel: clearLabel,
      DialogProps: DialogProps,
      okLabel: okLabel,
      todayLabel: todayLabel,
      cancelLabel: cancelLabel,
      DateInputProps: DateInputProps // @ts-ignore
      ,
      KeyboardDateInputComponent: KeyboardDateInputComponent // @ts-ignore
      ,
      PureDateInputComponent: PureDateInputComponent,
      wider: wider,
      showTabs: showTabs,
      displayStaticWrapperAs: displayStaticWrapperAs
    }, wrapperProps, restPropsForTextField), children);
  }

  return WrapperComponent;
}

function ownKeys$1(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread$1(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys$1(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys$1(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }
function makePickerWithStateAndWrapper(Wrapper, _ref) {
  var useInterceptProps = _ref.useInterceptProps,
      useValidation = _ref.useValidation,
      DefaultToolbarComponent = _ref.DefaultToolbarComponent;
  var PickerWrapper = makeWrapperComponent(Wrapper, {
    KeyboardDateInputComponent: KeyboardDateInput$1,
    PureDateInputComponent: PureDateInput
  });

  function PickerWithState(__props) {
    var utils = useUtils();
    var allProps = useInterceptProps(__props);
    var validationError = useValidation(allProps.value, allProps) !== null;

    var _usePickerState = usePickerState(allProps, {
      emptyValue: null,
      parseInput: parsePickerInputValue,
      areValuesEqual: function areValuesEqual(a, b) {
        return utils.isEqual(a, b);
      }
    }),
        pickerProps = _usePickerState.pickerProps,
        inputProps = _usePickerState.inputProps,
        wrapperProps = _usePickerState.wrapperProps; // Note that we are passing down all the value without spread.
    // It saves us >1kb gzip and make any prop available automatically on any level down.


    var value = allProps.value,
        onChange = allProps.onChange,
        other = _objectWithoutProperties(allProps, ["value", "onChange"]);

    var DateInputProps = _objectSpread$1({}, inputProps, {}, other, {
      validationError: validationError
    });

    return createElement(PickerWrapper, _extends({
      DateInputProps: DateInputProps,
      wrapperProps: wrapperProps
    }, other), createElement(Picker, _extends({}, pickerProps, {
      toolbarTitle: allProps.label || allProps.toolbarTitle,
      DateInputProps: DateInputProps,
      ToolbarComponent: other.ToolbarComponent || DefaultToolbarComponent
    }, other)));
  }

  var FinalPickerComponent = withDateAdapterProp(PickerWithState);
  return forwardRef(function (props, ref) {
    return createElement(FinalPickerComponent, _extends({}, props, {
      forwardedRef: ref
    }));
  });
}

export { DesktopWrapper as D, MobileWrapper as M, ResponsiveWrapper as R, StaticWrapper as S, makePickerWithStateAndWrapper as a, makeWrapperComponent as b, ResponsivePopperWrapper as c, DesktopPopperWrapper as d, makeValidationHook as m };
